### PR TITLE
config now supports data_dir

### PR DESCRIPTION
**Pull request by BruceBGordon**

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -20,14 +22,51 @@ function Config () {
...
 
+Config.prototype.getMode = function () {
+	if (argv.master) return 'master';
+	else if (argv.worker) return 'worker';
+	else throw new Error('Invalid mode.');
+}
+
+
+Config.prototype.expandTilde = function (apath) {
+	var homedir = process.env['HOME'];
```
Comment (thekushpatel) [inline lib/config.js:50]: Doesn't make sense for this function to be in Config. Config is just meant to be an abstraction over the config.json file. This is definitely a utility function.

File: lib/config.js [hunk 2]
```diff
@@ -20,14 +22,51 @@ function Config () {
 	for (var key in config) {
 		this[key] = config[key];
 	}
+	
+	if (mode === 'worker'){
```
Comment (thekushpatel) [inline lib/config.js]: Small comment, but worth mentioning for when I hook up JSHint.. Styling I've adopted for statements with brackets is adding a space like this `if (mode === worker) {`

File: lib/config.js [hunk 3]
```diff
@@ -20,14 +22,51 @@ function Config () {
...
+Config.prototype.expandTilde = function (apath) {
+	var homedir = process.env['HOME'];
+	if (!apath) return apath;
+	if (apath == '~') return homedir;
+	if (apath.slice(0, 2) != '~/') return apath;
+	return path.join(homedir, apath.slice(2));
+}
+
+Config.prototype.getGuerillaDir = function () {
+	return this.output_root_dir;
```
Comment (thekushpatel) [inline lib/config.js]: See previous comment about renaming this to `Config.prototype.getDataDir` and `return this.data_dir`.

File: lib/config.js [hunk 4]
```diff
@@ -36,7 +75,12 @@ Config.prototype.validate = function () {
 	if (!(urlObj.protocol && urlObj.port && urlObj.host && urlObj.hostname)) 
 		throw new Error('Hostname in config is invalid.');
 
-	var mode = utilities.getMode();
+	var mode = this.getMode();
+	if (mode === 'worker'){
+		if (!fs.lstatSync(this.output_root_dir).isDirectory()){
+			throw new Error('Config error. The supplied output_root_dir is not a directory:' + this.output_root_dir );
```
Comment (thekushpatel) [inline lib/config.js]: Bracket spacing.

File: package.json [hunk 5]
```diff
@@ -43,6 +43,7 @@
         "less-middleware": "^2.0.1",
         "moment": "^2.10.3",
         "morgan": "^1.5.1",
+        "mkdirp": "^0.5.1",
```
Comment (thekushpatel) [inline package.json]: No need for this extra dependency. We already have this functionality in one of our other libraries, fs-extra.

File: lib/config.js [hunk 6]
```diff
@@ -20,14 +22,51 @@ function Config () {
 	for (var key in config) {
 		this[key] = config[key];
 	}
+	
+	if (mode === 'worker'){
+		if (!this.output_root_dir) {
+			this.output_root_dir = '~/Guerilla';
+		}
+		this.output_root_dir = this.expandTilde(this.output_root_dir);
+		mkdirp.sync(this.output_root_dir);
```
Comment (thekushpatel) [inline lib/config.js]: Use this instead: https://github.com/jprichardson/node-fs-extra#mkdirsdir-callback

File: lib/config.js [hunk 7]
```diff
@@ -6,13 +6,15 @@
  * Reads and validates the master and worker config.json files.
  */
 
+var mkdirp = require("mkdirp");
+var fs = require('fs');
```
Comment (thekushpatel) [inline lib/config.js]: I think I always require `fs-extra` over `fs`.

File: config/worker/config.json.sample [hunk 8]
```diff
@@ -2,6 +2,7 @@
 	"name": "Guerilla Worker",
 	"host": "localhost",
 	"port": 8888,
+	"output_root_dir": "~/Guerilla",
```
Comment (thekushpatel) [inline config/worker/config.json.sample]: `data_dir`

File: lib/config.js (hunk 1 above)
Comment (BruceBGordon) [inline lib/config.js:50]: getMode is arguably part of the config abstraction. expandTilde is util*ish, but we cannot easily support circular dependencies. So we either move expandTilde to a util2 which doesn't depend on anything, or we leave this alone. For now it's ok to have expandTilde here as there are no other clients for it. GIven that we have data_dir most guerilla uses of  expandTilde can use data_dir

**Comment by thekushpatel** [comment]
Looks good.

**Comment by yahoocla** [comment]
CLA is valid!

**Pull request merged**